### PR TITLE
Separate learner model inputs from model state

### PR DESCRIPTION
**Pull request by RAMitchell**
## Summary

- rename `LearnerModelParam` to `LearnerModelState` to reflect that it is authoritative runtime/model state rather than a user parameter object
- move user inputs (`base_score`, `num_class`, `num_target`, and `boost_from_average`) into `LearnerTrainParam`
- materialize feature count, output shape, and intercept state when the learner first sees training data
- preserve empty learners in memory snapshots while requiring initialized state for ordinary prediction and model export
- remove `GradientBooster::ModelFitted()` as a proxy for learner initialization

The serialized `learner_model_param` key remains unchanged for model compatibility. Loading older snapshots also restores pending model inputs from that object when they are absent from `learner_train_param`.

This PR intentionally retains `need_configuration_` and the existing deferred component-configuration calls. It separates model initialization from that state machine first; configuration lifecycle cleanup can follow independently.

## Why

`LearnerModelParam` mixed user-provided construction inputs with model-derived, authoritative state. That made ordinary configuration responsible for creating a provisional model before training data was available and required booster internals to report whether the learner had been fitted. Separating the two lets configuration remain parameter handling while the first training operation creates a valid model state from both parameters and data.

## Validation

- `./build/testxgboost` (750 tests)
- focused learner/model-state and intercept tests after pre-commit formatting (11 tests)
- `pytest tests/python/test_basic_models.py tests/python/test_pickling.py tests/python/test_model_io.py tests/python/test_model_compatibility.py tests/python/test_multi_target.py` (61 passed, 1 skipped)
- `pre-commit run --files <34 changed files>`

### CODE REVIEW

**Comment by RAMitchell** [comment]
The point of this PR is to primarily to separate user supplied parameters from derived values that form the model. In the current code these two concepts are mixed together.

File: src/learner.cc [hunk 1]
```diff
@@ -346,144 +229,246 @@ std::string CanonicalizeBoosterName(std::string booster) {
...
+                                     tparam_.multi_strategy};
+    this->ValidateModelState();
+  }
+
+  [[nodiscard]] bst_feature_t InitNumFeatures(DMatrix const& train) const {
+    auto n_features = train.Info().num_col_;
+    error::MaxFeatureSize(n_features);
+    CHECK_NE(n_features, 0) << "0 feature is supplied. Are you using the raw Booster interface?";
+    return static_cast<bst_feature_t>(n_features);
+  }
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply src/learner.cc:344]: This should be a property of DMatrix to correctly compute n_columns.

**Review by trivialfis — COMMENTED** [review]
Skimmed through the PR, will take a deeper look later. Some questions:
- Which base score/intercept is serialized into the output JSON model? Before transformation by the objective or after?
- Since the state is not a user-facing parameter, and the original learner model param legacy is largely merged into the train param. This means the model state could be completely derived from the train param, maybe we should not serialize the model state (no need for the to json method).

**Comment by RAMitchell** [comment]
`base_score_value_` gets serialised in output space (I think with the exception of multi:softmax) and then gets converted to `base_score_` in additive margin space for runtime.

The model state is derived from both user params and the training dmatrix, so its unfortunately not safe to recreate.

**Review by trivialfis — APPROVED** [review]
